flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -331,9 +331,15 @@ class ExprLowering {
   genval(const Fortran::evaluate::ProcedureDesignator &proc) {
     if (const auto *intrinsic = proc.GetSpecificIntrinsic()) {
       auto signature = Fortran::lower::translateSignature(proc, converter);
+      // Intrinsic lowering is based on the generic name, so retrieve it here in
+      // case it is different from the specific name. The type of the specific
+      // is retained in the signature.
+      auto genericName =
+          converter.getFoldingContext().intrinsics().GetGenericIntrinsicName(
+              intrinsic->name);
       auto symbolRefAttr =
           Fortran::lower::getUnrestrictedIntrinsicSymbolRefAttr(
-              builder, getLoc(), intrinsic->name, signature);
+              builder, getLoc(), genericName, signature);
       mlir::Value funcPtr =
           builder.create<mlir::ConstantOp>(getLoc(), signature, symbolRefAttr);
```

Comment:
Should this be an `fir.addr_of` op? That's the way global symbols are referenced in other places.

I'm also curious if we'll need to support extended values here.

Reply:
I found that one had to use mlir::ConstantOp to create function pointers here: 
https://github.com/llvm/llvm-project/blob/master/mlir/include/mlir/Dialect/StandardOps/IR/Ops.td#L741
and here https://llvm.discourse.group/t/passing-function-pointer-as-argument-to-external-function/630.
Of course, that applies to mlir call_indirect, so maybe the situation should be different for FIR. I remember reading in some rational (that I cannot find back) that the use of ConstantOp to make a function pointer was related to mlir multithreading compilation, not 100% sure that applies in the FIR case and if my memory is correct here (I do not quite understand why it is related to multithereading).

Anyway, I tried `fir.addr_of` but, I am not sure how to get that working. I tried the follwing mlir with tco and that failed:
```
// No issue here with acos
func @__fs_acos_1(f32) -> f32 attributes {fir.runtime}
func @fir.acos.f32.ref_f32(%arg0: !fir.ref<f32>) -> f32 attributes {fir.intrinsic} {
  %0 = fir.load %arg0 : !fir.ref<f32>
  %1 = call @__fs_acos_1(%0) : (f32) -> f32
  return %1 : f32
}
                                                                                                      
func @test1() {                                                                                       
  %0 = fir.alloca f32                                                                                 
  %cst = constant 5.000000e-01 : f32                                                                  
  fir.store %cst to %0 : !fir.ref<f32>                                                                
  // issue here at fir level                                                                     
  %1 = fir.address_of(@fir.acos.f32.ref_f32) : (!fir.ref<f32>) -> f32                                 
  fir.call %1(%0) : (!fir.ref<f32>) -> f32                                                            
  return                                                                                              
}                                                                                                     
                                                                                                      
func @test2() {                                                                                       
  %0 = fir.alloca f32                                                                                 
  %cst = constant 5.000000e-01 : f32                                                                  
  fir.store %cst to %0 : !fir.ref<f32>                                                                
  // issue here at the llvm level                                                                                             
  %1 = fir.address_of(@fir.acos.f32.ref_f32) : !fir.ref<(!fir.ref<f32>) -> f32>                       
  %2 = fir.load %1 : !fir.ref<(!fir.ref<f32>) -> f32>                                                 
  fir.call %2(%0) : (!fir.ref<f32>) -> f32                                                            
  return                                                                                              
}  
```

test1 fails with:
> 'fir.address_of' op result #0 must be reference type, but got '(!fir.ref<f32>) -> f32'

test2 fails at llvm level with (if you comment test1):
> 'llvm.mlir.addressof' op the type must be a pointer to the type of the referenced function

Reply:
Thanks for the feedback and information. Maybe we don't really need `addr_of` any longer.